src/multiinstances/internal/ipc/ipcsocket.cpp

```diff
@@ -0,0 +1,209 @@
+{
+    if (!m_socket) {
+        m_lock = new IpcLock(serverName);
+        m_socket = new QLocalSocket();
+
+        QObject::connect(m_socket, &QLocalSocket::errorOccurred, [this](QLocalSocket::LocalSocketError err) {
+            //! NOTE If the server is down, then we will try to connect to another or create a server ourselves
+            if (err == QLocalSocket::PeerClosedError) {
+                m_disconected.notify();
+            }
+        });
+
+        QObject::connect(m_socket, &QLocalSocket::readyRead, [this]() {
+            onReadyRead();
+        });
+    }
+
+    if (m_socket->state() == QLocalSocket::ConnectedState) {
+        return true;
+    }
+
+    m_socket->connectToServer(serverName);
+    bool ok = m_socket->waitForConnected(500);
+    if (!ok) {
+        LOGW() << "failed connect to server";
+        return false;
+    }
+
+    m_socket->write(selfID().toUtf8());
+    ok = m_socket->waitForBytesWritten(500);
+    if (!ok) {
+        LOGE() << "failed init socket";
+    }
+
+    LOGI() << "success connected to ipc server";
+
+    return true;
+}
+
+bool IpcSocket::doSendToSocket(QLocalSocket* socket, const QByteArray& data)
+{
+    m_lock->lock();
+
+    socket->write(data);
+    bool ok = socket->waitForBytesWritten(500);
+    if (!ok) {
+        LOGE() << "failed write data to socket";
+        return false;
+    }
+
+    m_lock->unlock();
+
+    return true;
+}
+
+mu::async::Notification IpcSocket::disconected()
+{
+    return m_disconected;
+}
+
+bool IpcSocket::send(const Msg& msg)
+{
+    IF_ASSERT_FAILED(m_socket) {
+        return false;
+    }
+
+    Meta meta;
+    meta.id = selfID();
+
+    QByteArray data;
+    serialize(meta, msg, data);
+
+    IPCLOG() << data;
+    bool ok = doSendToSocket(m_socket, data);
+    if (!ok) {
+        LOGE() << "failed send to socket";
+    }
+    return ok;
+}
+
+void IpcSocket::onReadyRead()
+{
+    QByteArray data = m_socket->readAll();
+    IPCLOG() << data;
+    async::Async::call(this, [this, data]() {
+        onDataReceived(data);
+    });
+}
+
+void IpcSocket::onDataReceived(const QByteArray& data)
+{
+    IPCLOG() << "received: " << data;
+
+    Meta receivedMeta;
+    Msg receivedMsg;
+    deserialize(data, receivedMeta, receivedMsg);
+
+    if (receivedMsg.method.startsWith(IPC_)) {
+        onIpcMsg(receivedMeta, receivedMsg);
+        return;
+    }
+
+    if (receivedMeta.id == selfID()) {
+        return;
+    }
+
+    m_msgReceived.send(receivedMsg);
+}
+
+void IpcSocket::onIpcMsg(const Meta& receivedMeta, const Msg& receivedMsg)
+{
+    IPCLOG() << "received ipc msg: " << receivedMsg.method;
+
+    // answer on who is
+    if (receivedMsg.method == IPC_WHOIS) {
+        Msg answerMsg;
+        answerMsg.destID = SERVER_ID;
+        answerMsg.method = IPC_WHOIS;
+        answerMsg.args << selfID();
+        send(answerMsg);
+        return;
+    }
+
+    // receive meta info
+    if (receivedMsg.method == IPC_METAINFO) {
```

Comment:
could you explain the case when we'd receive this message type?

Reply:
This is one of the main service method.
It is sent when the number of instances or their state changes.
Thanks to him, all instances know how many more instances there are at the moment.